### PR TITLE
Modular Data Models: python adapter

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

this adds the `DataModel` interface that represents a modular data model module, an implementation of it for Python modules and a sample python module.

### Benefits

this will help building data model registry

### CODE REVIEW

File: irohad/ametsuchi/data_models/data_model_python.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  impl.func_commit_tx.release().dec_ref();
+  impl.func_commit_block.release().dec_ref();
+  impl.func_rollback_tx.release().dec_ref();
+  impl.func_rollback_block.release().dec_ref();
+  impl.python_module.release().dec_ref();
+  py::finalize_interpreter();
+}
+
+CommandResult DataModelPython::execute(
+    shared_model::interface::CallModel const &cmd) {
```
Comment (utsavjnn) [inline irohad/ametsuchi/data_models/data_model_python.cpp:85]: Shouldnt we use shared_model::proto::CallModel?

File: irohad/ametsuchi/data_models/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+set(PYTHON_EXECUTABLE:FILEPATH ${Python3_EXECUTABLE})
+
+find_package(pybind11 CONFIG REQUIRED)
+
+add_library(data_model_adapter_python data_model_python.cpp)
+
+target_compile_options(data_model_adapter_python
+  PRIVATE
+    -fsized-deallocation
+  )
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/CMakeLists.txt:13]:
```suggestion
if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
    target_compile_options(data_model_adapter_python PRIVATE -fsized-deallocation)
endif()
```

File: irohad/ametsuchi/data_models/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+find_package(Python3 REQUIRED COMPONENTS Interpreter)
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/CMakeLists.txt:1]:
```suggestion
# Copyright Soramitsu Co., Ltd. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

find_package(Python3 REQUIRED COMPONENTS Interpreter)
```

File: example/data_models/test_kv/kv_schema.proto [hunk 4]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
```
Comment (lebdron) [inline example/data_models/test_kv/kv_schema.proto:1]:
```suggestion
/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
 
syntax = "proto3";
```

File: example/data_models/test_kv/__init__.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import json
```
Comment (lebdron) [inline example/data_models/test_kv/__init__.py:1]:
```suggestion
#
# Copyright Soramitsu Co., Ltd. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
#

import json
```

File: irohad/ametsuchi/data_models/data_model_python.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  impl.func_commit_block.release().dec_ref();
+  impl.func_rollback_tx.release().dec_ref();
+  impl.func_rollback_block.release().dec_ref();
+  impl.python_module.release().dec_ref();
+  py::finalize_interpreter();
+}
+
+CommandResult DataModelPython::execute(
+    shared_model::proto::CallModel const &cmd) {
+  std::string cmd_str;  // = cmd.SerializeAsString();
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/data_model_python.cpp:86]:
```suggestion
  std::string const cmd_str{cmd.getTransport().SerializeAsString()};
```

File: irohad/ametsuchi/data_models/data_model_python.cpp [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+DataModelPython::~DataModelPython() {
+  Impl &impl = *impl_;
+  impl.func_execute.release().dec_ref();
+  impl.func_commit_tx.release().dec_ref();
+  impl.func_commit_block.release().dec_ref();
+  impl.func_rollback_tx.release().dec_ref();
+  impl.func_rollback_block.release().dec_ref();
+  impl.python_module.release().dec_ref();
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/data_model_python.cpp:80]:
```suggestion
  impl_.reset();
```

File: example/data_models/test_kv/__init__.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    cmd.ParseFromString(cmd_serialized)
+    print(cmd)
+    which = cmd.payload.WhichOneof('command')
+    global _tx_kv_storage
+    if which == 'set':
+        key = cmd.payload.set.key
+        val = cmd.payload.set.value
+        if not key in _tx_kv_storage:
+            if len(_tx_kv_storage) >= MAX_SIZE:
+                return (1, "storage limit exceeded")
```
Comment (lebdron) [inline example/data_models/test_kv/__init__.py:29]:
```suggestion
                return (3, "storage limit exceeded")
```

File: irohad/ametsuchi/data_models/data_model.hpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  class DataModel {
+   public:
+    virtual ~DataModel() = default;
+
+    virtual CommandResult execute(
+        shared_model::proto::CallModel const &cmd) = 0;
+
+    virtual void commit_transaction() = 0;
+
+    virtual void commit_block() = 0;
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/data_model.hpp]:
```suggestion
    virtual void commitBlock() = 0;
```

File: irohad/ametsuchi/data_models/data_model_python.hpp [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    // throws std::runtime_error
+    DataModelPython(std::vector<std::string> python_paths,
+                    std::string const &module_name,
+                    std::string const &initialization_argument);
+
+    ~DataModelPython();
+
+    CommandResult execute(shared_model::proto::CallModel const &cmd) override;
+
+    void commit_transaction() override;
```
Comment (lebdron) [inline irohad/ametsuchi/data_models/data_model_python.hpp]:
```suggestion
    void commitTransaction() override;
```
and so on

File: irohad/main/iroha_conf_loader.cpp [hunk 11]
```diff
@@ -470,6 +470,48 @@ inline void JsonDeserializerImpl::getVal<IrohadConfig::UtilityService>(
   getValByKey(path, dest.port, obj, config_members::Port);
 }
 
+template <>
+inline void JsonDeserializerImpl::getVal<IrohadConfig::DataModelModule::Python>(
+    const std::string &path,
+    IrohadConfig::DataModelModule::Python &dest,
+    const rapidjson::Value &src) {
+  assert_fatal(src.IsObject(),
+               path + " utility service config top element must be an object.");
```
Comment (lebdron) [inline irohad/main/iroha_conf_loader.cpp:479]:
```suggestion
               path + " must be an object.");
```

File: shared_model/interfaces/common_objects/impl/data_model_id.cpp [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+using namespace shared_model::interface;
+
+bool DataModelId::operator==(DataModelId const &rhs) const {
+  return name == rhs.name and version == rhs.version;
+}
+
+std::size_t DataModelId::Hasher::operator()(DataModelId const &id) const {
+  std::hash<std::string> hasher;
+  return hasher(id.name) + hasher(id.version);
```
Comment (lebdron) [inline shared_model/interfaces/common_objects/impl/data_model_id.cpp:19]:
```suggestion
  // TODO rework without addition
  return hasher(id.name) + hasher(id.version);
```

**Pull request merged**